A68-random-subsetting.md

```diff
@@ -0,0 +1,117 @@
+* The policy tries to distribute connections among servers as equally as possible. The higher `(N_clients*subset_size)/N_servers` ratio is, the closer the resulting server connection distribution is to uniform.
+* The policy minimizes the amount of connection churn generated during server scale-ups by using [rendezvous hashing](https://en.wikipedia.org/wiki/Rendezvous_hashing)
+
+### Subsetting algorithm
+
+* The policy receives a single configuration parameter: `subset_size`, which must be configured by the user.
+* When the lb policy is initialized it also creates a random 32-byte long `salt` string. 
+* After every resolver update the policy picks a new subset. It does this by implementing `rendezvous hashing` algorithm:
+  * Concatenate string representation of every address in the list with `salt`.
+  * For every resulting entity compute [MurmurHash3](https://en.wikipedia.org/wiki/MurmurHash) hash, which produces 128-byte output.
+  * Sort all entities by hash.
+  * Pick first `subset_size` values from the list.
+* Pass the resulting subset to the child LB policy.
+* If the number of addresses is less than `subset_size` always use all available addresses.
+
+### Characteristics of the selected algorithm
+
+#### Uniform connection distribution on high scale
+
+When `(N_clients*subset_size)/N_servers` ratio is high, the resulting connection distribution between servers is close to uniform. This is because the chosen hash function is uniform and every server has equal probability to be chosen by every client. 
+Though it could be done, we don't provide any mathematical guaranties about the resulting connection distribution. Any such guarantees will be probabilistic and have limited value in practice. Instead, we can give you a few samples:
+* N_clients = 100, N_servers = 100, subset_size = 5
+![State Diagram](A68_graphics/subsetting100-100-5.png)
+* N_clients = 100, N_servers = 100, subset_size = 25
+![State Diagram](A68_graphics/subsetting100-100-25.png)
+* N_clients = 100, N_servers = 10, subset_size = 5
+![State Diagram](A68_graphics/subsetting100-10-5.png)
+* N_clients = 500, N_servers = 10, subset_size = 5
+![State Diagram](A68_graphics/subsetting500-10-5.png)
+* N_clients = 2000, N_servers = 10, subset_size = 5
+![State Diagram](A68_graphics/subsetting2000-10-5.png)
+
+#### Low connection churn during server rollouts
+
+The graphs provided in the previous section prove this is the case in practice (we rollout all servers in the middle of every test) Low connection churn during server rollouts is the primary motivation why rendezvous hashing was used as the subsetting algorithm: it guaranties that if a single server is either added or removed to the IP address list, every client will update at most 1 entry in its subset. This is because the hashes for all unaffected servers remain stable, which guarantees that the order of the servers after sorting also remains stable. 
```

Comment:
> It might be worth showing how other implementations perform with regards to connection churn. It could also be useful to explain to the reader why the previous graphs demonstrate low connection churn as it might not be obvious at first sight.

updated the doc to include more details on that.

> I'm also wondering what is the trade-off in situations where more than one server gets added or removed from the list? Is it still superior to use rendezvous hashing or would there be a better algorithm? Not disagreeing with the decision here but trying to cover all grounds

The same logic applies: the order of unaffected servers remains stable and number of client-side updates is always less or equal to the number of updated servers. Also mentioned this in the doc.

Reply:
> Regarding connection churn, we also get additional chaos across clients due to the fact that we are hashing on IPs rather than some coordinated id, right? That is, if a server we are talking to gets rescheduled to a different pod/cluster we could reshuffle unnecessarily.

That is correct. gRPC has consistent hashing LB policy for such cases. @atollena has an in-flight gRFC related to that https://github.com/grpc/proposal/pull/412

Reply:
My understanding of the consistent hashing lb work is that it's solving the traditional problem of mapping a key to a server. 

I was referring to the fact that rendezvous hashing in general is stable so long as the object we hash on are stable. In this case, we are hashing on IPs, which are relatively stable but probably add some degree of churn to our ordering when pods get rescheduled. This isn't a big deal given that we are doing random subsets to begin with, but just wanted to verify my understanding.